From e4e294e2c69f813e9696a67737b5df045862c350 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hmehrtens@maxlinear.com>
Date: Thu, 5 Aug 2021 14:26:51 +0200
Subject: bridge: Add option to configure snooping on a bridge

This option allows to configure the netlink snooping option on bridges.
---
 Makefile.am                              |  1 +
 include/netlink/route/link/bridge_info.h | 29 ++++++++++
 lib/route/link/bridge_info.c             | 68 +++++++++++++++++++++++-
 libnl-route-3.sym                        |  2 +
 4 files changed, 99 insertions(+), 1 deletion(-)
 create mode 100644 include/netlink/route/link/bridge_info.h

diff --git a/Makefile.am b/Makefile.am
index fcfdd69..6283458 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -148,6 +148,7 @@ libnlinclude_netlink_route_linkdir = $(libnlincludedir)/netlink/route/link
 libnlinclude_netlink_route_link_HEADERS = \
 	include/netlink/route/link/api.h \
 	include/netlink/route/link/bonding.h \
+	include/netlink/route/link/bridge_info.h \
 	include/netlink/route/link/bridge.h \
 	include/netlink/route/link/can.h \
 	include/netlink/route/link/geneve.h \
diff --git a/include/netlink/route/link/bridge_info.h b/include/netlink/route/link/bridge_info.h
new file mode 100644
index 0000000..8c9739d
--- /dev/null
+++ b/include/netlink/route/link/bridge_info.h
@@ -0,0 +1,29 @@
+/*
+ * netlink/route/link/bridge_info.h		Bridge
+ *
+ *	This library is free software; you can redistribute it and/or
+ *	modify it under the terms of the GNU Lesser General Public
+ *	License as published by the Free Software Foundation version 2.1
+ *	of the License.
+ *
+ * Copyright (c) 2021 MaxLinear, Inc.
+ */
+
+#ifndef NETLINK_LINK_BRIDGE_INFO_H_
+#define NETLINK_LINK_BRIDGE_INFO_H_
+
+#include <netlink/netlink.h>
+#include <netlink/route/link.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+extern int rtnl_link_bridge_set_mcast_snooping(struct rtnl_link *link, uint8_t activate);
+extern int rtnl_link_bridge_get_mcast_snooping(struct rtnl_link *link, uint8_t *activate);
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif
diff --git a/lib/route/link/bridge_info.c b/lib/route/link/bridge_info.c
index b4265eb..4df680d 100644
--- a/lib/route/link/bridge_info.c
+++ b/lib/route/link/bridge_info.c
@@ -18,15 +18,19 @@
  */
 
 #include <netlink-private/netlink.h>
+#include <netlink/route/link/bridge_info.h>
 #include <netlink-private/route/link/api.h>
 
+#define BRIDGE_ATTR_MCAST_SNOOPING	(1 << 0)
+
 struct bridge_info
 {
+	uint8_t			b_mcast_snooping;
 	uint32_t		ce_mask; /* to support attr macros */
 };
 
 static struct nla_policy bi_attrs_policy[IFLA_BR_MAX+1] = {
-
+	[IFLA_BR_MCAST_SNOOPING]	= { .type = NLA_U8 },
 };
 
 static inline struct bridge_info *bridge_info(struct rtnl_link *link)
@@ -68,6 +72,11 @@ static int bridge_info_parse(struct rtnl_link *link, struct nlattr *data,
 
 	bi = link->l_info;
 
+	if (tb[IFLA_BR_MCAST_SNOOPING]) {
+		bi->b_mcast_snooping = nla_get_u8(tb[IFLA_BR_MCAST_SNOOPING]);
+		bi->ce_mask |= BRIDGE_ATTR_MCAST_SNOOPING;
+	}
+
 errout:
 	return err;
 }
@@ -81,8 +90,16 @@ static int bridge_info_put_attrs(struct nl_msg *msg, struct rtnl_link *link)
 	if (!data)
 		return -NLE_MSGSIZE;
 
+	if (bi->ce_mask & BRIDGE_ATTR_MCAST_SNOOPING)
+		NLA_PUT_U8(msg, IFLA_BR_MCAST_SNOOPING,
+			   bi->b_mcast_snooping);
+
 	nla_nest_end(msg, data);
 	return 0;
+
+nla_put_failure:
+	nla_nest_cancel(msg, data);
+	return -NLE_MSGSIZE;
 }
 
 static void bridge_info_free(struct rtnl_link *link)
@@ -105,6 +122,55 @@ static struct rtnl_link_info_ops bridge_info_ops = {
 		return -NLE_OPNOTSUPP; \
 	}
 
+/**
+ * Set MC Src IP Lookup
+ * @arg link		Link object of type bridge
+ * @arg ip_lookup	IP Lookup to set.
+ *
+ * @see rtnl_link_bridge_get_mc_dest_ip_lookup()
+ *
+ * @return Zero on success, otherwise a negative error code.
+ * @retval -NLE_OPNOTSUPP Link is not a bridge
+ */
+int rtnl_link_bridge_set_mcast_snooping(struct rtnl_link *link, uint8_t activate)
+{
+	struct bridge_info *bi = bridge_info(link);
+
+	IS_BRIDGE_INFO_ASSERT(link);
+
+	bi->b_mcast_snooping = activate;
+	bi->ce_mask |= BRIDGE_ATTR_MCAST_SNOOPING;
+
+	return 0;
+}
+
+/**
+ * Get MC Dest IP Lookup
+ * @arg link		Link object of type bridge
+ * @arg ip_lookup	Output argument.
+ *
+ * @see rtnl_link_bridge_set_mc_dest_ip_lookup()
+ *
+ * @return Zero on success, otherwise a negative error code.
+ * @retval -NLE_OPNOTSUPP Link is not a bridge
+ */
+int rtnl_link_bridge_get_mcast_snooping(struct rtnl_link *link, uint8_t *activate)
+{
+	struct bridge_info *bi = bridge_info(link);
+
+	IS_BRIDGE_INFO_ASSERT(link);
+
+	if (!(bi->ce_mask & BRIDGE_ATTR_MCAST_SNOOPING))
+		return -NLE_NOATTR;
+
+	if (!activate)
+		return -NLE_INVAL;
+
+	*activate = bi->b_mcast_snooping;
+	return 0;
+}
+
+
 static void __init bridge_info_init(void)
 {
 	rtnl_link_register_info(&bridge_info_ops);
diff --git a/libnl-route-3.sym b/libnl-route-3.sym
index 7cc5a3d..a52ca23 100644
--- a/libnl-route-3.sym
+++ b/libnl-route-3.sym
@@ -270,6 +270,8 @@ global:
 	rtnl_link_bridge_set_cost;
 	rtnl_link_bridge_set_flags;
 	rtnl_link_bridge_set_learning_limit;
+	rtnl_link_bridge_set_mcast_snooping;
+	rtnl_link_bridge_get_mcast_snooping;
 	rtnl_link_bridge_set_port_state;
 	rtnl_link_bridge_set_priority;
 	rtnl_link_bridge_str2flags;
